pkg/component/controller/workerconfig/reconciler_test.go

```diff
@@ -0,0 +1,756 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workerconfig
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/k0sproject/k0s/internal/testutil"
+	k0sv1beta1 "github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/v1beta1"
+	"github.com/k0sproject/k0s/pkg/component/controller/leaderelector"
+	"github.com/k0sproject/k0s/pkg/constant"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kubeletv1beta1 "k8s.io/kubelet/config/v1beta1"
+	"k8s.io/utils/pointer"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"sigs.k8s.io/yaml"
+)
+
+type (
+	obj           = map[string]any
+	arr           = []any
+	kubeletConfig = kubeletv1beta1.KubeletConfiguration
+)
+
+func TestMain(m *testing.M) {
+	logrus.StandardLogger().SetLevel(logrus.DebugLevel)
+	m.Run()
+}
+
+func TestReconciler_Lifecycle(t *testing.T) {
+	cleaner := newMockCleaner()
+	underTest, err := NewReconciler(
+		constant.GetConfig(t.TempDir()),
+		&k0sv1beta1.ClusterSpec{
+			API: &k0sv1beta1.APISpec{},
+			Network: &k0sv1beta1.Network{
+				ClusterDomain: "test.local",
+				ServiceCIDR:   "99.99.99.0/24",
+			},
+		},
+		testutil.NewFakeClientFactory(),
+		&leaderelector.Dummy{Leader: true},
+	)
+	require.NoError(t, err)
+	underTest.cleaner = cleaner
+
+	t.Run("fails_to_start_without_init", func(t *testing.T) {
+		err := underTest.Start(context.TODO())
+		require.Error(t, err)
+		require.Equal(t, "cannot start: workerconfig.reconcilerCreated", err.Error())
+	})
+
+	t.Run("init", func(t *testing.T) {
+		assert.NoError(t, underTest.Init(context.TODO()))
+		assert.Equal(t, "init", cleaner.state.Load())
+	})
+
+	t.Run("another_init_fails", func(t *testing.T) {
+		err := underTest.Init(context.TODO())
+		require.Error(t, err)
+		assert.Equal(t, "cannot initialize: workerconfig.reconcilerInitialized", err.Error())
+		assert.Equal(t, "init", cleaner.state.Load())
+	})
+
+	mockApplier := installMockApplier(t, underTest)
+
+	stopTest := func() func(t *testing.T) {
+		var once sync.Once
+		return func(t *testing.T) {
+			once.Do(func() {
+				assert.NoError(t, underTest.Stop())
+				assert.Equal(t, "stop", cleaner.state.Load())
+			})
+		}
+	}()
+
+	t.Run("starts", func(runT *testing.T) {
+		require.NoError(runT, underTest.Start(context.TODO()))
+		t.Cleanup(func() { stopTest(t) })
```

Comment:
It would be called when the whole test will be finished, not just this subtest.

Reply:
That was the intention. What about this?

```diff
diff --git a/pkg/component/controller/workerconfig/reconciler_test.go b/pkg/component/controller/workerconfig/reconciler_test.go
--- a/pkg/component/controller/workerconfig/reconciler_test.go
+++ b/pkg/component/controller/workerconfig/reconciler_test.go
@@ -94,8 +94,9 @@ func TestReconciler_Lifecycle(t *testing.T) {
 		}
 	}()
 
-	t.Run("starts", func(runT *testing.T) {
-		require.NoError(runT, underTest.Start(context.TODO()))
+	t.Run("starts", func(startT *testing.T) {
+		require.NoError(startT, underTest.Start(context.TODO()))
+		// Ensure the component gets stopped when the parent test exits
 		t.Cleanup(func() { stopTest(t) })
 	})

Reply:
It's still unclear, why we call it here and what it's supposed to do. Again, back to AAA, what is the assert here?

Reply:
The arrangement has been done by the previous subtests. The action is `underTest.Start(context.TODO())`. The assertion is that the component could be started, i.e. the Start method didn't return an error. The cleanup function on the parent test is to ensure that the (now started) component will be stopped when the lifecycle test finishes.

The lifecycle test ensures that the reconciler adheres to the component lifecycle. Each subtest builds upon the previous ones. Maybe it helps to make it clear what's going on when looking at the test structure only:

```
--- PASS: TestReconciler_Lifecycle (0.00s)
    --- PASS: TestReconciler_Lifecycle/fails_to_start_without_init (0.00s)
    --- PASS: TestReconciler_Lifecycle/init (0.00s)
    --- PASS: TestReconciler_Lifecycle/another_init_fails (0.00s)
    --- PASS: TestReconciler_Lifecycle/starts (0.00s)
    --- PASS: TestReconciler_Lifecycle/another_start_fails (0.00s)
    --- PASS: TestReconciler_Lifecycle/reconciles (0.00s)
    --- PASS: TestReconciler_Lifecycle/stops (0.00s)
    --- PASS: TestReconciler_Lifecycle/stop_may_be_called_again (0.00s)
    --- PASS: TestReconciler_Lifecycle/reinit_fails (0.00s)
    --- PASS: TestReconciler_Lifecycle/restart_fails (0.00s)
```

In this case the subtests are not independent of each other, but more like consecutive steps. An alternative would be to replace the subtests with some block comments and additional message params for the assertions.

Reply:
> Each subtest builds upon the previous ones

> In this case the subtests are not independent of each other, but more like consecutive steps

That's exactly the problem. Tests should be independent and self-sufficient. Anyone should be able to run the single subtest without any trouble.

Reply:
Done.